File: installer/support/env_setup/dotenv.ts
#!/usr/bin/env -S deno run --allow-all --no-lock
import { $, $$ } from "../dax.ts"

import { dimosEnvVars } from "../constants.ts"
import { addGitIgnorePatterns, getProjectDirectory } from "../misc.ts"
import * as p from "../prompt_tools.ts"

export async function setupDotenv(projectPath: string, envPath: string): Promise<boolean> {
    let envExists = false
    try {
        const st = await Deno.stat(envPath)
        envExists = st.isFile
    } catch {
        envExists = false
    }

    if (!envExists) {
        console.log(`Dimos involves setting several project specific environment variables.\nWe highly recommend having these in a git-ignored ${p.highlight(`.env`)} file.\n`)
        if (!p.askYesNo(`I don't see a ${p.highlight(`.env`)} file, can I create one for you?`)) {
            console.log("- Okay, I'll assume you will manage env vars yourself:")
            for (const [name, value] of Object.entries(dimosEnvVars)) {
                console.log(`  ${name}=${value}`)
            }
            // skip rest of phase (no .envrc if they won't even make a .env)
            return false
        }
        await Deno.writeTextFile(envPath, "")
        await addGitIgnorePatterns(projectPath, ["/.env"], { comment: "Added by dimos setup" })
    }

    let envText = ""
    try {
        envText = await Deno.readTextFile(envPath)
    } catch {
        envText = ""
    }
    const existingVarsInDotEnv = new Set(
        envText
            .split("\n")
            .map((line) => line.trim())
            .map((line) => line.match(/^(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=/)?.[1])
            .filter((name) => !!name),
    )

    const missingEnvVars: string[] = []
    for (const [name, value] of Object.entries(dimosEnvVars)) {
        if (!existingVarsInDotEnv.has(name)) {
            missingEnvVars.push(`${name}=${value}`)
        }
    }

    if (missingEnvVars.length > 0) {
        const needsTrailingNewline = envText.length > 0 && !envText.endsWith("\n")
        const additions = (needsTrailingNewline ? "\n" : "") + missingEnvVars.join("\n") + "\n"
        await Deno.writeTextFile(envPath, envText + additions)
        p.boringLog(`- appended ${missingEnvVars.length} env var(s) to .env`)
    } else {
        p.boringLog("- all required env vars already exist in .env")
    }

    return true
}

File: installer/pyz_app/support/env_setup/dotenv.py
#!/usr/bin/env python3
from __future__ import annotations

import re
from pathlib import Path

from ..constants import dimosEnvVars
from ..misc import add_git_ignore_patterns
from .. import prompt_tools as p


def setup_dotenv(project_path: str | Path, env_path: str | Path) -> bool:
    project_path = Path(project_path)
    env_path = Path(env_path)

    env_exists = env_path.is_file()

    if not env_exists:
        print(
            "Dimos involves setting several project specific environment variables.\n"
            f"We highly recommend having these in a git-ignored {p.highlight('.env')} file.\n"
        )
        if not p.ask_yes_no(f"I don't see a {p.highlight('.env')} file, can I create one for you?"):
            print("- Okay, I'll assume you will manage env vars yourself:")
            for name, value in dimosEnvVars.items():
                print(f"  {name}={value}")
            return False
        env_path.write_text("")
        add_git_ignore_patterns(project_path, ["/.env"], {"comment": "Added by dimos setup"})

    try:
        env_text = env_path.read_text()
    except FileNotFoundError:
        env_text = ""

    existing_vars = {
        match.group(1)
        for line in env_text.split("\n")
        if (match := re.match(r"^(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=", line.strip()))
    }

    missing_env_vars = [f"{name}={value}" for name, value in dimosEnvVars.items() if name not in existing_vars]

    if missing_env_vars:
        needs_trailing_newline = len(env_text) > 0 and not env_text.endswith("\n")
        additions = ("\n" if needs_trailing_newline else "") + "\n".join(missing_env_vars) + "\n"
        env_path.write_text(env_text + additions)
        p.boring_log(f"- appended {len(missing_env_vars)} env var(s) to .env")
    else:
        p.boring_log("- all required env vars already exist in .env")

    return True


__all__ = ["setup_dotenv"]
